test/BasicAccessTest.php

```diff
@@ -20,70 +20,95 @@
 
 class BasicAccessTest extends TestCase
 {
-    /** @var ServerRequestInterface|ObjectProphecy */
+    /** @var ObjectProphecy */
     private $request;
 
-    /** @var UserRepositoryInterface|ObjectProphecy */
+    /** @var ObjectProphecy */
     private $userRepository;
 
-    /** @var UserInterface|ObjectProphecy */
+    /** @var ObjectProphecy */
     private $authenticatedUser;
 
-    /** @var ResponseInterface|ObjectProphecy */
+    /** @var ObjectProphecy */
     private $responsePrototype;
 
     /** @var callable */
     private $responseFactory;
 
+    /**
+     * @return void
+     */
     protected function setUp()
     {
         $this->request = $this->prophesize(ServerRequestInterface::class);
         $this->userRepository = $this->prophesize(UserRepositoryInterface::class);
         $this->authenticatedUser = $this->prophesize(UserInterface::class);
         $this->responsePrototype = $this->prophesize(ResponseInterface::class);
-        $this->responseFactory = function () {
+
+        $this->responseFactory = $this->responseFactoryClosure();
+    }
+
+    /**
+     * @return callable(): object
+     */
+    private function responseFactoryClosure(): callable
+    {
+        return function () {
+            /** @var object */
             return $this->responsePrototype->reveal();
         };
     }
 
-    public function testConstructor()
+    public function testConstructor(): void
     {
+        /** @var UserRepositoryInterface $userRepositoryInterface */
```

Comment:
There should be no need to annotate this inline as when you re-add property annotations properly, psalm should already know whats being revealed.

Reply:
Would be very nice if you could suggest here something.
I've no clue why psalm cannot understand mixed annotations properly except one reason:

Too old PHPUnit version. Fo correct understanding mocks psalm using [an additional plugin](https://github.com/psalm/psalm-plugin-phpunit) and in stubs directory, I see there is Prophecy.phpstub

But it looks like Prophecy.phpstub isn't suitable for PHPUnit 7.5

Reply:
The type inference works with internal phpunit mocks, but not  with prophecy, so the same feedback as above stands: IMO we should just phase out prophecy.